src/Microsoft.SqlTools.ServiceLayer/LanguageServices/BindingQueue.cs

```diff
@@ -286,86 +296,119 @@ private void ProcessQueue()
                             continue;
                         }
 
-                        bool lockTaken = false;
-                        try
-                        {                                                    
-                            // prefer the queue item binding item, otherwise use the context default timeout
-                            int bindTimeout = queueItem.BindingTimeout ?? bindingContext.BindingTimeout;
+                        var bindingContextTask = this.BindingContextTasks[bindingContext];
 
-                            // handle the case a previous binding operation is still running                                                 
-                            if (!bindingContext.BindingLock.WaitOne(queueItem.WaitForLockTimeout ?? 0))
-                            {
-                                queueItem.Result = queueItem.TimeoutOperation != null
-                                    ? queueItem.TimeoutOperation(bindingContext)
-                                    : null;
+                        // Run in the binding context task in case this task has to wait for a previous binding operation
+                        this.BindingContextTasks[bindingContext] = bindingContextTask.ContinueWith((task) =>
+                        {
+                            bool lockTaken = false;
+                            try
+                            {                                                    
+                                // prefer the queue item binding item, otherwise use the context default timeout
+                                int bindTimeout = queueItem.BindingTimeout ?? bindingContext.BindingTimeout;
+
+                                // handle the case a previous binding operation is still running
+                                if (!bindingContext.BindingLock.WaitOne(queueItem.WaitForLockTimeout ?? 0))
+                                {
+                                    try
+                                    {
+                                        Logger.Write(TraceEventType.Warning, "Binding queue operation timed out waiting for previous operation to finish");
```

Comment:
Consider defining a new Logger.LogEvent entry and using that as the argument to Logger.Write. This will help in future to quickly search for all log entries with a given LogEvent.

The LogEvents could be specific areas of code that we designate them to be.